# **Linear Regression**

## **Importing Libraries**

In [13]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

## **Loading Dataset**

In [14]:
data=fetch_california_housing(as_frame=True)
df=data.frame
df

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422
...,...,...,...,...,...,...,...,...,...
20635,1.5603,25.0,5.045455,1.133333,845.0,2.560606,39.48,-121.09,0.781
20636,2.5568,18.0,6.114035,1.315789,356.0,3.122807,39.49,-121.21,0.771
20637,1.7000,17.0,5.205543,1.120092,1007.0,2.325635,39.43,-121.22,0.923
20638,1.8672,18.0,5.329513,1.171920,741.0,2.123209,39.43,-121.32,0.847


In [15]:
df.shape

(20640, 9)

## **Data Preprocessing**|

### **Missing Values**

In [16]:
df.isnull().sum()

MedInc         0
HouseAge       0
AveRooms       0
AveBedrms      0
Population     0
AveOccup       0
Latitude       0
Longitude      0
MedHouseVal    0
dtype: int64

There are no missing values

## **Defining Features & target Variable**

In [17]:
x=df.drop('MedHouseVal',axis=1)
y=df['MedHouseVal']

### Splitting Data into Training & Test sets

In [18]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

## **Feature Scaling for Regression**

In [19]:
scaler=StandardScaler()
x_train_scaled=scaler.fit_transform(x_train)
x_test_scaled=scaler.transform(x_test)

Scaling ensures all features are on a similar range/scale, which helps regression models converge faster and perform better.

* scaler = StandardScaler()

  1. Creates a StandardScaler object from sklearn.preprocessing.

  2. StandardScaler standardizes features by removing the mean and scaling to unit variance.

  3. Important for regression algorithms (like linear regression, SVR) to perform well.

* X_train_scaled = scaler.fit_transform(X_train)

  1. fit_transform() fits the scaler to X_train (calculates mean & standard deviation) and transforms the training data.

  2. Produces X_train_scaled, a scaled version of X_train.

* X_test_scaled = scaler.transform(X_test)

  1. Uses the same scaling parameters (mean & std from training set) to scale X_test.

  2. Ensures that the test data is transformed consistently with training data.

## **Linear Regression Model**

In [20]:
#Model Training
model=LinearRegression()
model.fit(x_train_scaled,y_train)

#Prediction
y_pred=model.predict(x_test_scaled)

## **Model Evaluation**



In [21]:
print("\nMean Squared error :", mean_squared_error(y_test, y_pred))
print("R² Score:",r2_score(y_test,y_pred))


Mean Squared error : 0.5558915986952442
R² Score: 0.575787706032451


* mean_squared_error(y_test, y_pred)

  1. Calculates the Mean Squared Error (MSE) between actual values (y_test) and predicted values (y_pred).

  2. Measures the average squared difference between predicted and true values.

  3. Lower MSE → better model performance.

* r2_score(y_test, y_pred)

  1. Calculates the R² score (coefficient of determination).

  2. Measures how well the model explains the variance in the target variable.

  3. R² ranges from 0 to 1; closer to 1 → better fit.

* Why not accuracy, precision, etc.?

  1. Accuracy, precision, recall are for classification problems (categorical targets).

  2. Here, MedHouseVal is continuous (regression problem), so we use regression metrics like MSE and R².

✅ Summary:

Regression → MSE, R²

Classification → Accuracy, Precision, Recall



* Mean Squared Error (MSE) = 0.556

  1. On average, the squared difference between predicted and actual house values is 0.556.

  2. A higher MSE indicates the **model’s predictions are not very close to actual values**.

* R² Score = 0.576

  1. The model explains only about 57.6% of the variance in the target variable.

  2. This is a **moderate fit, not very strong**.

* Interpretation:

   **Linear Regression is not performing very well** on this dataset.

   To improve performance, we can try a more flexible model like Random Forest Regressor, which can capture non-linear relationships better.

## **Polynomial Features**

In [27]:
from sklearn.preprocessing import PolynomialFeatures

poly=PolynomialFeatures(degree=2,include_bias=False)
x_train_poly=poly.fit_transform(x_train_scaled)
x_test_poly=poly.transform(x_test_scaled)

print("Original Features : \n",x_train_scaled.shape[1])
print("Polynomial features :\n",x_train_poly.shape[1])

# Linear Regression on Polynomial Features
lr_poly = LinearRegression()
lr_poly.fit(x_train_poly, y_train)

# Predictions
y_pred_train = lr_poly.predict(x_train_poly)
y_pred_test = lr_poly.predict(x_test_poly)

# Evaluation
r2_train = r2_score(y_train, y_pred_train)
r2_test = r2_score(y_test, y_pred_test)
print("R2_Score :\n",r2_test)

print("Mean squared Error :\n",mean_squared_error(y_test,y_pred_test))

Original Features : 
 8
Polynomial features :
 44
R2_Score :
 0.6456819729261623
Mean squared Error :
 0.46430152383015905
